File: backend/app/schemas/round.py
from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


def _validate_bid_trick_range(v: int) -> int:
    if not (0 <= v <= 13):
        raise ValueError("each value must be between 0 and 13")
    return v


class RoundCreate(BaseModel):
    """Schema for creating a round (bids)"""

    bids: List[int] = Field(..., min_length=4, max_length=4, description="List of 4 bids (0-13)")
    trump_suit: Optional[str] = Field(
        None, description="Trump suit: spades, clubs, diamonds, hearts, no-trump"
    )

    @field_validator("bids", mode="after")
    @classmethod
    def bids_in_range(cls, v: List[int]) -> List[int]:
        for bid in v:
            _validate_bid_trick_range(bid)
        return v


class TricksSubmit(BaseModel):
    """Schema for submitting tricks"""

    tricks: List[int] = Field(
        ..., min_length=4, max_length=4, description="List of 4 tricks taken (0-13)"
    )
    bids: List[int] = Field(
        ...,
        min_length=4,
        max_length=4,
        description="List of 4 bids (0-13) - from previous submission",
    )
    trump_suit: Optional[str] = Field(
        None, description="Trump suit: spades, clubs, diamonds, hearts, no-trump"
    )

    @field_validator("tricks", "bids", mode="after")
    @classmethod
    def values_in_range(cls, v: List[int]) -> List[int]:
        for x in v:
            _validate_bid_trick_range(x)
        return v


class RoundResponse(BaseModel):
    """Schema for round response"""

    id: int
    game_id: UUID
    round_number: int
    bids: List[int]
    tricks: List[int]
    scores: List[int]
    round_mode: str
    trump_suit: Optional[str] = None
    created_by: Optional[UUID] = None
    created_at: datetime

    class Config:
        from_attributes = True  # For SQLAlchemy model conversion

File: flutter_mobile/lib/models/round.dart
/// Round from API; aligned with backend RoundResponse.
class Round {
  final int id;
  final String gameId;
  final int roundNumber;
  final List<int> bids;
  final List<int> tricks;
  final List<int> scores;
  final String roundMode;
  final String? trumpSuit;
  final String? createdBy;
  final DateTime createdAt;

  const Round({
    required this.id,
    required this.gameId,
    required this.roundNumber,
    required this.bids,
    required this.tricks,
    required this.scores,
    required this.roundMode,
    this.trumpSuit,
    this.createdBy,
    required this.createdAt,
  });

  factory Round.fromJson(Map<String, dynamic> json) {
    return Round(
      id: json['id'] as int,
      gameId: json['game_id'] as String,
      roundNumber: json['round_number'] as int,
      bids: List<int>.from(
          (json['bids'] as List).map((e) => e is int ? e : int.parse('$e'))),
      tricks: List<int>.from(
          (json['tricks'] as List).map((e) => e is int ? e : int.parse('$e'))),
      scores: List<int>.from(
          (json['scores'] as List).map((e) => e is int ? e : int.parse('$e'))),
      roundMode: json['round_mode'] as String? ?? 'over',
      trumpSuit: json['trump_suit'] as String?,
      createdBy: json['created_by']?.toString(),
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}

/// Request: submit bids (backend RoundCreate).
class RoundCreate {
  RoundCreate({required this.bids, this.trumpSuit});

  final List<int> bids;
  final String? trumpSuit;

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{'bids': bids};
    if (trumpSuit != null) map['trump_suit'] = trumpSuit;
    return map;
  }
}

/// Request: submit tricks (backend TricksSubmit).
class TricksSubmit {
  TricksSubmit({
    required this.tricks,
    required this.bids,
    this.trumpSuit,
  });

  final List<int> tricks;
  final List<int> bids;
  final String? trumpSuit;

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      'tricks': tricks,
      'bids': bids,
    };
    if (trumpSuit != null) map['trump_suit'] = trumpSuit;
    return map;
  }
}
